README.md
# pyGUI
python3 modular GUI based on PyQt4

Hier möchte ich ein Projekt vorstellen, dass im [Fachgebiet Messtechnik](http://www.uni-kassel.de/eecs/fachgebiete/messtechnik/) der [Universität Kassel](http://www.uni-kassel.de/) entstanden ist.<br>
Im Rahmen einer Dissertation entstand als Nebenprodukt eine Software-Umgebung um schnell und einfach Python Programme in eine GUI einzubetten und Messungen oder Auswertungen der Messdaten durchführen zu können. Die Wahl der Programmiersprache fiel auf Python aus mehreren Gründen:
* große Anzahl an Bibliotheken zur Signalverarbeitung und gute Darstellungsmöglichkeiten
* hervorragende Unterstützung von QT als GUI Backend
* Möglichkeiten der Einbindung externer Bibliotheken und der Wegfall der teueren Lizenzkosten, wie es bei MATLAB (wurde hauptsächlich früher im Fachgebiet eingesetzt) der Fall ist.

<br>Ich möchte aber ausdrücklich betonen, dass die Veröffentlichung der pyGUI als Opensource-Lizenz von mir und Stanislav Tereschenko ausschließlich privat erfolgt. **Das heißt, dass die Veröffentlichung vom Fachgebiet Messtechnik gestattet wurde, aber das Fachgebiet Messtechnik keinerlei Support oder Haftung übernimmt!**<br><br>
Darüber hinaus möchte ich um Verständnis werben, dass aus verständlichen Gründen nur die GUI an sich und keinerlei Messabläufe, Algorithmen oder Hardwareschnittstellen in der pyGUI enthalten sind.

### Installing
* install winPython 3.4.3.x or python3-spyderlib
```
$ git clone https://github.com/pgollor/pyGUI.git
$ cd pyGUI
$ mkdir modules
$ cp -R ./sampleModules/* ./modules/
$ chmod 755 main.py
```
For more information please see the german description below or the doxygen help in ```doc/index.chm``` folder.<br>

### Doxygen help as html
```
$ doxygen pyGUI.doxyfile
```

### Usage
```
$ ./main.py
```
For debugging use the ```-d``` option and for logging ```-l LOGLEVEL```.<br>
You can use ```--help``` for detail information.

### [Change log](CHANGELOG.md)

### Todo
* replace all old qt signals with new qt signal management
* add linux help
* use other plot functions


## Grundidee
Im Fachgebiet sollten verschiedene Messabläufe insbesondere für hochauflösende Interferometer aber auch anschließende Auswertung und Darstellung der Messergebnisse in Python realisiert werden. Das kann man entweder mit X verschiedenen Messprogramm realisieren, oder man entwickelt ein modulares Konstrukt, das man mit den verschiedensten Modulen füttern kann, um Messungen durchführen zu können und eine grafische Ausgabe daraus zu generieren. Daraus entstand die Grundidee zur pyGUI.<br><br>

Aus dieser Grundidee ist eine kleine Eierlegendewollmilchsau geworden, die wir gerne der Opensourcegemeinschaft zur Verfügung stellen möchten. Leider ist aus Zeitgründen die Dokumentation noch sehr rudimentär bzw. schnell zusammen geschrieben. Das werde ich aber so schnell wie möglich versuchen nachzubessern. Aktuell wird mit doxygen eine pdf und eine Windows-Hilfe erstellt. Die Windows-Hilfe soll unter Windows auch noch Zeitnah in die pyGUI direkt eingebunden werden.

## Grobes Konstrukt
Da das Programm modular aufgebaut ist, ist es nicht auf Messabläufe beschränkt. Jeder erdenklicher Programmablauf lässt sich in verschiedenen Modulen nachbilden, mit der die pyGUI gefüttert werden kann.<br>
Das heißt, dass die pyGUI eine grafische Benutzeroberfläche zur Verfügung stellt die mit Qt realisiert wurde und die Module auf einige Standardelemente (Logging, GUI Elemente automatisch speichern, …) zurückgreifen können.<br>
Dazu gibt es drei verschiedene arten von Modulen die man in der pyGUI hinzufügen kann:
* *hidden* Module
* *settings* Module
* *application* Module

Das *hidden* Modul eignet sich perfekt um eine Schnittstelle zu einer DLL, Hardware oder sonst etwas für andere Module herzustellen. Auf diese Schnittstelle kann dann zum Beispiel ein *settings* Modul oder ein *applikation* Modul zugreifen. Das *hidden* Modul hat dabei keine grafische Ausgabe.<br>
Das *settings* und das *application* Modul sind im Grunde genommen gleichwertige Module mit der Ausnahme, dass die *apllication* Module in der GUI direkt angezeigt werden und die *settings* Module nur über das Menü „Module Settings“ zu erreichen sind. Dies soll die bekannt Struktur von Programmen widerspiegeln.<br><br>

Ein paar Testmodule sind in der GUI unter „sampleModules“ enthalten und können bei bedarf in den Ordner „modules“ eingefügt werden. Zum aktivieren des Moduls muss dieses unter dem Menüreiter „Main->Modules“ ausgewählt werden. Nach einem Neustart des Programms wird es automatisch geladen.

## Voraussetzung
Um pyGUI benutzen zu können wird eine python3 Umgebung mit qt und ein paar anderen Bibliotheken benötigt. Am einfachsten Benutzt man unter Windows das winPython Paket und unter Linux die python3-spyderlib. Dann einfach die main.py starten und los geht es.<br><br>

pyGUI kann sowohl unter Windows 7 und 8 als auch unter Linux verwendet werden. Bisher ist der Einsatz aber nur unter Windows getestet worden. Bei anderen Oberflächen wie Gnome oder KDE kann es zu einer anderen Darstellung kommen. Da würden wir uns über zahlreiche Rückmeldungen freuen. Dazu und zu allen anderen Themen der pyGUI kann gerne auf meiner Homepage diskutiert werden.


# License

Dieses Projekt ist lizensiert als Inhalt der
Creative Commons Namensnennung - Weitergabe unter gleichen Bedingungen 3.0 Unported-Lizenz.<br>
Um eine Kopie der Lizenz zu sehen, besuchen Sie http://creativecommons.org/licenses/by-sa/3.0/.<br>
-- englisch version --<br>
This work is licensed under the Creative Commons Attribution-ShareAlike 3.0 Germany License.<br>
To view a copy of this license, visit http://creativecommons.org/licenses/by-sa/3.0/ or send a letter to<br>
Creative Commons, 444 Castro Street, Suite 900, Mountain View, California, 94041, USA.


CHANGELOG.md
Changes
=======

## 2015-07-09
* fix bug with logarithm slider signals
* add warning for smal screens

## 2015-06-29
* use dock widgets insteat of mdiArea
* split better slider into seperated classes
* add some more documentation
* renew sample modules
* move dock widgets

## 2015-06-04
* move moduele settings from ```modulename```.xml to ```user_settings.ini```
* use qsettings instead of xml settings management
* replace some old qt signals with new signal usage
* for windows: add help function direcly into the GUI
* use python parser to activate debuging with ```-d``` option


main.py
#! /usr/bin/python3

##
# @file main.py
# 
# @date 25.07.2013
# @author Stanislav Tereschenko
# @author Pascal Gollor (http://www.pgollor.de)
# 
# @copyright Dieses Projekt ist lizensiert als Inhalt der
# Creative Commons Namensnennung - Weitergabe unter gleichen Bedingungen 3.0 Unported-Lizenz.<br>
# Um eine Kopie der Lizenz zu sehen, besuchen Sie http://creativecommons.org/licenses/by-sa/3.0/.<br>
# -- englisch version --<br>
# This work is licensed under the Creative Commons Attribution-ShareAlike 3.0 Germany License.<br>
# To view a copy of this license, visit http://creativecommons.org/licenses/by-sa/3.0/ or send a letter to<br>
# Creative Commons, 444 Castro Street, Suite 900, Mountain View, California, 94041, USA.
# 
# @mainpage pyGUI
# These project is a universal python3 GUI. You can add your own modules and use some global functions for your project.
#  
# @section modules Module
# @subsection main Allgemein
# Es gibt eine abstrakte Oberklasse(abstractModuleClass) von dieser alle Module abgeleitet werden.
# Drei verschiedene Modulklassen sind moeglich:
# - hidden
# - settings
# - application
#
# Die einzelnen Module werden in dieser Reihenfolge geladen.
# Innerhalb eines Modultyps werden die Module nach alphabethischer Reihenfolge geladen.
# Falls es notwendig ist, dass ein Modul zwingend von einem anderen des gleichen Typs abhaengt, sollte man den Ordner des Moduls mit einer Zahl vor dem Modulnahmen versehen.
# Falls Modul xy zwingend das Modul zz benoetigt, sollte (NUR) der Ordnernamen des Moduls zz z.B. in 10zz umbenannt werden.
# Damit ist sicher gestellt, dass das Modul zz vor dem Modul xy geladen wird.
# 
# Jedes Modul muss ein eigenen Unterordner mit dem Namen des Modules im Ordner modules besitzen.
# In diesem Ordner muessen folgende Dateien mindestens enthalten sein:
# - [MODULNAME].py
# - [MODULNAME].ui (nicht fuer hidden Module) 
# 
# Die genaue Struktur dieser Dateien kann in den Beispielmodulen nachgelesen werden.
# Die Klasse des Moduls muss immer module heissen und von der jeweils dazugehoerigen Elternklasse abgeleitet sein: 
# - class module(applicationModuleClass): 
# - class module(applicationModuleClass):
# - class module(hiddenModuleClass):
# 
# @subsection settings Einstellungen laden und speichern
# Fuer die Module "settings" und "application" koennen Werte verschiedener Qt Elemente automatisch gespeichert werden.
# Dazu fragt de rModulhandler die Funktion "getDefaultSettings" in jedem Modul ab.
# Falls die Funktion nicht vorhanden ist, wird davon ausgegangen, dass keine Einstellungen gespeichert werden sollen.
# Als Rueckgabewerte muss ein Dictionary zurueck gegeben werden in dem die zu speichernden Informationen enthalten sind.
# Um z.B. den Inhalt eines QLineEdits zu speichern saehe das dict wie folgt aus:  
# <pre>
#{
#"inhaltXY" = {"qName": "labelXY", "value" = "Irgend ein Inhalt"}
#}</pre>
# 
# Die beiden Parameter "qName" und "value" muessen fuer jeden Eintrag zwingend vorhanden ein.
# Der Schluessel "inhaltXY" kann hingegen frei gewaehlt werden.
# Dies entspricht danna auch dem Listeneintrag wenn das Modul waehrend der Laufzeit seine Einstellungen mit "self._getSettings()" abfragt.
# @n Der Wert zu "qName" ist der Name des dazugehoerigen Qt Elements und "value" der Inhalt der als default angegeben wird.
# Das heisst je nach Qt Element kann "value" auch leer sein.
# @n Neben QLabel werden von der GUI auch noch weitere Qt Elemente unterstuetzt:
# - QCheckBox, QRadioButton und QGroupBox:
# @n Es wird der Status "checked()" gespeichert. Dazu ist "value" als bool anzugeben. Die Groupbox muss dafuer "checkable" sein.
# <pre>{
#  "X" = {"qName": "checkBoxX", "value": False},
#  "Y" = {"qName": "radioButtonY", "value": True},
#  "Z" = {"qName": "groupBoxZ", "value": True}
#}</pre>
# - QComboBox
# <pre>{
#  "X" = {"qName": "comboBoxX", "value": "Inhalt"}
#}</pre>
# - QSpinBox und QDoubleSpinBox
# @n Beid er spinBox sollten Integers verwendet werden und fuer die doubleSpingBox floats.
# <pre>{
#  "X" = {"qName": "spinBoxX", "value": 2, "minVal": 0, "maxVal": 10, "step": 1},
#  "Y" = {"qName": "doubleSpinBoxY", "value": 1.0, "minVal": 0.0, "maxVal": 10.0, "step": 0.1},
#}</pre>
# - customIntegerSlider, customFloatSlider, customLogarithmSlider
# @n vohanden sein muessen; "qName", "value", "maxVal" und "minVal"
# @n optiional sind: "step" (default: 1), "pageStep" (default 10), "connectedLabels", "connectedLineEdits"
# <pre>{
#  "X": {"qName": "horizontalSliderX", "type": "double", "value": 10, "minVal": 0, "maxVal": 100, "step": 0.1, "pageStep": 20, "connectedLabels": ["labelTestSl"], "connectedLineEdits": ["lineEditTextSl"]},
#}</pre>
# - QTextEdit und QPlainTextEdit
# <pre>{
#  "X" = {"qName": "textEdit", "value": "Inhalt"}
#}</pre>
# 
# Falls in der Gui unter Settings die Einstellung Save on exit aktiviert ist, werden dann die geaenderten Einstellung automatisch beim Beenden des Programms gespeichert und nach erneutem Start wieder geladen.
# 
# @section Modultypen
# @subsection hidden-Modul
# Ein hidden Modul enthaelt keine Funktionen fuer die eigentliche Oberflaeche. In diesen Modulen werden nur Funktionen oder externe Klassen bereit gestellt die von anderen Modulen verwendet werden. 
# 
# @subsection settings-Modul
# Diese Module werden in dem Tab "settings" angezeigt und setzen z.B. einfach nur Einstellungen fuer andere Module (im Idealfall der hidden Module).
# Aber diese stellen schon eine graphische Oberflaeche fuer die Einstellungen durch den Benutzer bereit. 
# 
# @subsection application-Modul
# Diese Module sind die eigentliche Benutzeroberflaeche. Fuer jedes Modul wird ein eigener Tab geladen.
# In diesem Tab werden automatisch die Oberflaechenelemente von Qt geladen, die in der dazu gehoerigen ui-Datei gespeichert sind. 
#
# @section function Funktionen in der Modulklasse
# @subsection must-have must have
# Die Funktion die auf jeden Fall in der Modulklasse vorhanden sein muss ist die init Funktion:
# <pre>
# class module(applicationModuleClass):
#   def __init__(self, parent, name):
#     applicationModuleClass.__init__(self, parent, name)
#
#     self.setDisplayName('Modulname')
#   # end __init__
# </pre>
# 
# @subsection optional
# Alle anderen Funktionen sind als Virtuelle-Funktionen bereits vorhanden die ueberladen werden koennen.
# Diese werden in folgender Reihenfolge aufgerufen:
# - initPreSettings(self): Wird geladen bevor die Einstellungen aus der XML-Datei geladen werden.
# - initModule(self): Ist zum initialisieren von Klassen aus anderen Modulen und fuer Klassen aus dem eigenen Ordner gedacht. Dazu wird fuer diese Funktion vom Modulhandler in das Modulverzeichnis gewechselt.
# - initGui(self): Ist zum initialisieren der GUI Elemente gedacht.
# - onClose(self): Wird beim Beenden aufgerufen bevor die Klasse geloescht wird.
# - onActive(self): Wird jedes mal fuer das Modul aufgerufen, dass in der GUI ausgewaehlt wurde.
# - onInactive(self): Gegenteil von "onActive".
# - getDefaultSettings(self)
#
# Die Funktionen initPreSettings und initModule werden hintereinander nach der oben beschriebenen Modulreihenfolge ausgefuehrt.
# Nach dem diese Funktionen bei ALLEN Modulen ausgefuehrt wurden, wird nach der gleichen Reihenfolge die Funktion initGui ausgefuehrt.
# 


import sys, os, logging, optparse
from PyQt4 import QtGui
from mainWindow import mainWindow
from functions.guiLogger import QtLogger

# add dlls folder to environment
dll_path = os.path.join(os.path.abspath(os.path.dirname(__file__)), "dlls")
openblas_path = os.path.dirname(sys.executable) + "\\..\\tools\\mingw32\\bin"
user_dlls = os.pathsep.join((dll_path, openblas_path))
global_path = os.environ.get('PATH', '')

if dll_path not in global_path:
	os.environ['PATH'] = os.pathsep.join((user_dlls, global_path))
# end if


def parser():
	# parser
	parser = optparse.OptionParser(
		usage = "%prog [options]",
		description = "pyGUI - Modular GUI based on python and qt."
	)
	
	group = optparse.OptionGroup(parser, "Diagnostics")
	group.add_option("-d", "--debug",
		dest = "debug",
		action = "store_true",
		help = "Show debug output. And override loglevel with debug.",
		default = False
	)
	group.add_option("-l", "--loglevel",
		dest = "loglevel",
		action = "store",
		type = 'int',
		help = "Loglevel for python logger. " + str(logging.CRITICAL) + ": critical  " + str(logging.ERROR) + ": error  " + str(logging.WARNING) + ": warning  " + str(logging.INFO) + ":info  " + str(logging.DEBUG) + ":debug",
		default = logging.ERROR
	)
	parser.add_option_group(group)

	(options, args) = parser.parse_args()
	
	if (options.debug == True):
		options.loglevel = logging.DEBUG
	# end if
	
	return options
# end parser



## @brief main function
# create main window
if __name__ == '__main__':
	## set basic logging settings add own logger stream
	# Set loglevel not to debug, because qt create much debug outputs. 
	logging.basicConfig(level = logging.WARNING, format='%(asctime)-9s %(name)s [%(levelname)s]: %(message)s', datefmt='%H:%M:%S', handlers = [QtLogger()])
	
	# parser
	options = parser()

	# create Qt application
	app = QtGui.QApplication(sys.argv)
	
	if (options.debug == True):
		w = mainWindow(loglevel = options.loglevel, debug = options.debug)
		w.init(app.desktop().availableGeometry())
		w.show()
	else:
		try:
			w = mainWindow(loglevel = options.loglevel, debug = options.debug)
		
			# init main window
			w.init(app.desktop().availableGeometry())
	
			# show main gui	
			w.show()
		except Exception as e:
			message = "unresolved exception in main.py: "

			if (len(e.args) > 0):
				message += str(e.args)
			# end if

			print(message)
			sys.exit(0)
		# end try
	# end if
	
	#app.setQuitOnLastWindowClosed(True)
	ret = app.exec_()
	
	# delete all objects
	del(w)
	del(app)

	os._exit(0) # kill abrupt
	sys.exit(ret)
# end if
